server/polar/email_update/service.py

```diff
@@ -0,0 +1,130 @@
+import datetime
+from math import ceil
+from urllib.parse import urlencode
+
+from sqlalchemy import delete
+from sqlalchemy.orm import joinedload
+
+from polar.auth.models import AuthSubject
+from polar.config import settings
+from polar.email.renderer import get_email_renderer
+from polar.email.sender import get_email_sender
+from polar.exceptions import PolarError
+from polar.kit.crypto import generate_token_hash_pair, get_token_hash
+from polar.kit.extensions.sqlalchemy import sql
+from polar.kit.services import ResourceServiceReader
+from polar.kit.utils import utc_now
+from polar.models import EmailVerification
+from polar.models.user import User
+from polar.postgres import AsyncSession
+
+from .schemas import EmailUpdateCreate
+
+TOKEN_PREFIX = "polar_ev_"
+
+
+class EmailUpdateError(PolarError): ...
+
+
+class InvalidEmailUpdate(EmailUpdateError):
+    def __init__(self) -> None:
+        super().__init__(
+            "This email update request is invalid or has expired.", status_code=401
+        )
+
+
+class EmailUpdateService(ResourceServiceReader[EmailVerification]):
+    async def request_email_update(
+        self,
+        email: str,
+        session: AsyncSession,
+        auth_subject: AuthSubject[User],
+        expire_at: datetime.datetime | None = None,
```

Comment:
This is not useful, the expiration date is always generated by default on the model itself.